cdap-data-fabric/src/main/java/co/cask/cdap/data2/transaction/RetryingTransactionSystemClient.java

```diff
@@ -0,0 +1,135 @@
+/*
+ * Copyright © 2017 Cask Data, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not
+ * use this file except in compliance with the License. You may obtain a copy of
+ * the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+package co.cask.cdap.data2.transaction;
+
+import co.cask.cdap.api.Predicate;
+import co.cask.cdap.common.service.Retries;
+import co.cask.cdap.common.service.RetryStrategy;
+import com.google.common.base.Supplier;
+import org.apache.tephra.InvalidTruncateTimeException;
+import org.apache.tephra.Transaction;
+import org.apache.tephra.TransactionCouldNotTakeSnapshotException;
+import org.apache.tephra.TransactionNotInProgressException;
+import org.apache.tephra.TransactionSystemClient;
+import org.apache.thrift.TException;
+
+import java.io.InputStream;
+import java.util.Collection;
+import java.util.Set;
+
+/**
+ * Base class that helps in retrying certain calls to the transaction system.
+ * Unfortunately, this class is tightly coupled with the TransactionServiceClient in Tephra,
+ * as it determines a retryable failure by checking if the TransactionSystemClient threw a RuntimeException wrapping
+ * a TException, which is what TransactionServiceClient does. Once Tephra implements better exceptions, this class
+ * can be removed.
+ *
+ * By default, it doesn't retry anything. Subclasses should override the methods that should be retried.
+ */
+public abstract class RetryingTransactionSystemClient implements TransactionSystemClient {
+  private static final Predicate<Throwable> IS_RETRYABLE = new Predicate<Throwable>() {
+    @Override
+    public boolean apply(Throwable input) {
+      return input instanceof RuntimeException && input.getCause() != null && input.getCause() instanceof TException;
```

Comment:
Looking at the Tephra code, it seems to be using TException to decide when to apply it's own retry strategy (https://github.com/apache/incubator-tephra/blob/master/tephra-core/src/main/java/org/apache/tephra/distributed/TransactionServiceClient.java#L209). If so, it means Tephra has a bug or I've missed some place in the code.